src/backends/plonky2/recursion/circuit.rs

```diff
@@ -0,0 +1,627 @@
+        witness::{PartialWitness, WitnessWrite},
+    },
+    plonk::{
+        circuit_builder::CircuitBuilder,
+        circuit_data::{
+            CircuitConfig, CircuitData, ProverCircuitData, VerifierCircuitData,
+            VerifierCircuitTarget,
+        },
+        proof::{ProofWithPublicInputs, ProofWithPublicInputsTarget},
+    },
+    recursion::dummy_circuit::cyclic_base_proof,
+};
+
+use crate::{
+    backends::plonky2::{
+        basetypes::{Proof, C, D},
+        error::{Error, Result},
+    },
+    middleware::F,
+};
+
+/// InnerCircuit is the trait used to define the logic of the circuit that is
+/// computed inside the RecursiveCircuit.
+pub trait InnerCircuit: Clone {
+    type Input;
+
+    fn build(
+        builder: &mut CircuitBuilder<F, D>,
+        // public_inputs will be used for the recursive proof verification
+        public_inputs: Vec<Target>,
+        selectors: Vec<BoolTarget>,
+    ) -> Result<Self>;
+
+    /// assigns the values to the targets
+    fn set_targets(&self, pw: &mut PartialWitness<F>, input: &Self::Input) -> Result<()>;
+}
+
+#[derive(Clone, Debug)]
+pub struct RecursiveParams {
+    /// determines the arity of the RecursiveCircuit
+    arity: usize,
+}
+
+/// RecursiveCircuit defines the circuit that verifies `arity` proofs of itself.
+pub struct RecursiveCircuit<I: InnerCircuit> {
+    params: RecursiveParams,
+    selectors_targ: Vec<BoolTarget>,
+    innercircuit_targ: I,
+    proofs_targ: Vec<ProofWithPublicInputsTarget<D>>,
+    // cyclic-recursion params:
+    verifier_data_targ: VerifierCircuitTarget,
+    verifier_data: VerifierCircuitData<F, C, D>,
+}
+
+impl<I: InnerCircuit> RecursiveCircuit<I> {
+    pub fn build(
+        builder: &mut CircuitBuilder<F, D>,
+        params: &RecursiveParams,
+        // self's verifier_data
+        verifier_data: VerifierCircuitData<F, C, D>,
+    ) -> Result<Self> {
+        let selectors_targ: Vec<BoolTarget> = (0..params.arity)
+            .map(|_| builder.add_virtual_bool_target_safe())
+            .collect();
+
+        // proof verification
+        let common_data = verifier_data.common.clone();
+        let verifier_data_targ = builder.add_verifier_data_public_inputs();
+        let proofs_targ: Result<Vec<ProofWithPublicInputsTarget<D>>> = (0..params.arity)
+            .map(|i| {
+                let proof_targ = builder.add_virtual_proof_with_pis(&common_data);
+                builder.conditionally_verify_cyclic_proof_or_dummy::<C>(
+                    selectors_targ[i],
+                    &proof_targ,
+                    &common_data,
+                )?;
+                Ok(proof_targ)
+            })
+            .collect();
+        let proofs_targ = proofs_targ?;
+
+        // get the targets of the public inputs used to verify the proofs, to be
+        // used in the InnerCircuit
+        let pub_inp: Vec<Target> = proofs_targ
+            .iter()
+            .flat_map(|proof_pis| proof_pis.public_inputs.clone())
```

Comment:
These public inputs are built like this:
[inner_circuit_public, circuit_digest, constants_sigmas_cap].
We should do two things here:
1. If the `selectors_targ=true`, then `circuit_digest, constants_sigmas_cap` from the proof should be equal to the values in `verifier_data_targ`.  Or maybe we skip the conditional part and just connect, beacuse the dummy proof may use the same verifier data?
2. Extract the `inner_circuit_public` and that's collected and passed to `I::build`

Reply:
>     1. If the `selectors_targ=true`, then `circuit_digest, constants_sigmas_cap` from the proof should be equal to the values in `verifier_data_targ`.  Or maybe we skip the conditional part and just connect, beacuse the dummy proof may use the same verifier data?

I learned that this is already being done through the combination of
```rust
builder.add_verifier_data_public_inputs()
```
and
```rust
builder.conditionally_verify_cyclic_proof_or_dummy()
```
https://github.com/0xPolygonZero/plonky2/blob/6a2c1b47b7b00864c2b65a44c18099b0152f550e/plonky2/src/recursion/cyclic_recursion.rs#L114-L117

Reply:
yes, this is this way